cdap-security/src/main/java/io/cdap/cdap/security/authorization/AuthorizerInstantiator.java

```diff
@@ -103,65 +101,70 @@ public AuthorizerInstantiator(CConfiguration cConf, AuthorizationContextFactory
    * authorization is disabled.
    */
   @Override
-  public synchronized Authorizer get() {
-    if (authorizer != null) {
-      return authorizer;
-    }
-
-    if (!authorizationEnabled) {
+  public Authorizer get() {
+    if (!cConf.getBoolean(Constants.Security.Authorization.ENABLED)) {
       LOG.debug("Authorization is disabled. Authorization can be enabled  by setting " +
                   Constants.Security.Authorization.ENABLED + " to true.");
-      authorizer = new NoOpAuthorizer();
-      return authorizer;
+      return NOOP_AUTHORIZER;
     }
-    if (!authenticationEnabled) {
+    if (!cConf.getBoolean(Constants.Security.ENABLED)) {
       LOG.info("Authorization is enabled. However, authentication is disabled. Authorization policies will not be " +
                  "enforced. To enforce authorization policies please enable both authorization, by setting " +
                  Constants.Security.Authorization.ENABLED + " to true and authentication, by setting " +
                  Constants.Security.ENABLED + "to true.");
-      authorizer = new NoOpAuthorizer();
+      return NOOP_AUTHORIZER;
+    }
+
+    // Authorization is enabled
+    Authorizer authorizer = this.authorizer;
+    if (authorizer != null) {
       return authorizer;
     }
-    // Authorization is enabled, so continue with startup now
-    String authorizerExtensionJarPath = cConf.get(Constants.Security.Authorization.EXTENSION_JAR_PATH);
-    String authorizerExtraClasspath = cConf.get(Constants.Security.Authorization.EXTENSION_EXTRA_CLASSPATH);
-    if (Strings.isNullOrEmpty(authorizerExtensionJarPath)) {
-      throw new IllegalArgumentException(
-        String.format("Authorizer extension jar path not found in configuration. Please set %s in cdap-site.xml to " +
-                        "the fully qualified path of the jar file to use as the authorization backend.",
-                      Constants.Security.Authorization.EXTENSION_JAR_PATH));
+
+    synchronized (this) {
+      authorizer = this.authorizer;
```

Comment:
This approach would still allow double creation if two threads enter here at same time. Not sure if that's fine for you. An alternative is to use `com.google.common.base.Suppliers#memoize`

Reply:
No it doesn't. The authorizer is a singleton with the use of volatile.

Reply:
Yep, you are correct, sorry. We doublecheck within synchronized